### DISCUSSION TITLE
Volume issue with docker

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#15986] [Archive]
Hi .. I used this guide to start an instance of directus in docker: https://docs.directus.io/self-hosted/installation/docker.html

This is my experimental command where I mount the volumes:
```
docker run -d --restart=always -p 8026:8055 -v /srv/directus/:/directus --name d9 -e KEY=255d861b-5ea1-5996-9aa3-922530ec40b1 -e SECRET=6116487b-cda1-52c2-b5b5-c8022c45e263 -e ADMIN_EMAIL="test@mail.com" -e ADMIN_PASSWORD="secret" directus/directus
``` 
After that I have a running directus and I can log in and creating everything. BUT in my folder /srv/directus there are only three folder created without any content. If I go inside the docker in the folder /directus/database is a sqlite database created. 

What is wrong?

**Comment by br41nslug** [comment]
I think you can't mount the entire `/directus` folder, and if you could you shouldn't as then you can't update Directus anymore because all files would be persisted in the volume 😬

**Reply by ghost** [reply]
I switched for my easy start to the cloud. 

But I need to persist the data and otherwise I loose them during an docker move or rebuild. And I tried to mount only the subdirectory of database with the same result.

This is written in the doc;

Persistence
Containers are ephemeral, and this means that whenever you stop a container, all the data associated with it is going to be removed [unless you persist them](https://docs.docker.com/storage) when creating your container.

Directus image by default [will use the following locations](https://github.com/directus/directus/blob/main/docker/Dockerfile#L56-L60) for data persistence (note that these can be changed through environment variables)

/directus/uploads for uploads
/directus/database (only when using SQLite and not configured to a different folder)
/directus/extensions for loading extensions

**Reply by azrikahar** [reply]
@br41nslug is right that we can't mount the entire `/directus` folder directly. Although it does understandably feels like the simpler approach compared to the recommended `/uploads`, `/database` and `/extensions` directory in the docs, one pitfall here is we will also persist the `node_modules` directory and the `package.json` file, thus effectively making it not possible to "update" Directus anymore without wiping the relevant files. Here's a quick `ls -l` of `/directus` for a better depiction said issue:

```
drwxr-xr-x    2 node     node          4096 Oct 13 11:15 database
drwxr-xr-x   10 node     node          4096 Oct 13 08:59 extensions
drwxr-xr-x  710 node     node         24576 Sep 22 21:21 node_modules
-rw-r--r--    1 node     node          1064 Sep 22 21:17 package.json
drwxr-xr-x    2 node     node          4096 Sep 22 21:21 uploads
```

That being said, can you try `-v /srv/directus/database:/directus/database` instead? Make sure not to add trailing slash for the host mount on the left side.

**Reply by ghost** [reply]
Okay .. now I understand the problem .. if I mount only the database .. the folder are created correctly

07:43:21 ✨ Initializing bootstrap...
07:43:46 🚨 Can't connect to the database.
07:43:46 🚨 SQLITE_CANTOPEN: unable to open database file
Error: SQLITE_CANTOPEN: unable to open database file
07:43:50 ✨ Initializing bootstrap...

**Reply by azrikahar** [reply]
Judging from the SQLITE_CANTOPEN error, I think this is a file permission/docker mount issue and not technically an issue with Directus. 

Can you try checking the container's `node` user UID and GID by running `docker exec d9 id`[^1], then compare it with the host's sqlite file's user UID and GID. You will need to make sure the host's sqlite file permission matches the container's `node` user. 

[^1]: I'm using `d9` as you used `--name d9` in your original command, make sure to update it if it's not `d9` anymore

**Reply by ghost** [reply]
This is the result ...

uid=1000(node) gid=1000(node) groups=1000(node)

**Reply by ghost** [reply]
And you are completly right .. I set the permissions of the folder database from 755 to 777 and the database was created .. I need to mount the three volumes seperatly and set the permissions to 777 .. now it´s working.

**Reply by azrikahar** [reply]
Glad it works! 

Setting `777` might be perceived as too permissive, so I think you should also be able to resolve this by using chown 1000:1000 (node user's UID and GID) for the files on the host. Similar to this solution shared by a community member: https://github.com/directus/directus/discussions/6480#discussioncomment-917708

Alternatively you might be able to extend the Directus docker image to change the `node` user UID and GID to match the host user as described in the docker-node docs (Directus docker image is based on this image): https://github.com/nodejs/docker-node/blob/main/docs/BestPractices.md#non-root-user, specifically this section:

> If you need to change the uid/gid of the user you can use:
>
> `RUN groupmod -g 999 node && usermod -u 999 -g 999 node`

**Reply by crediblebytes** [reply]
Is there a way to get this working on synology NAS? I am having the same error and passing in the user/group **env vars** is how I've gotten every other container to work but it doesn't seem to have an effect. 
PUID: 1026
PGID: 100
